### ISSUE TITLE
Restoration visits reserved for internal use?

### ISSUE THREAD

**Issue by gregblass** [turbolinks/turbolinks#77]
Per the documentation: "Applications using the iOS or Android adapters initiate a restoration visit when moving backward in the navigation stack. Restoration visits have an action of restore and Turbolinks reserves them for internal use. You should not attempt to annotate links or invoke Turbolinks.visit with an action of restore."

In my case, I haven't rolled out native wrappers quite yet (I plan on it in the future). For my responsive mobile web app though, I'm essentially doing the same thing - I'm replacing my hamburger icon with a back button when the user navigates from an index to a show, for example. I'm hard coding back links throughout the app, which works fine (show back link goes back to index) - until the user starts coming from another place, like an activity feed - then the back button is navigating them to a place they didn't come from.

As far as I can tell, I would need the restore action in this case, like:

    <a data-turbolinks-action="restore">Back</a>

And saving the scroll position of the page would be great in the case of an activity feed, or for other long pages my app is based around.

I'm sure there was a reason to not allow restoration visits - just wondering what it was? If we're using back buttons on mobile web in the same manner that you would in a native app, what would be the recommended practice to restore the previous page from the navigation stack?

**Comment by sstephenson** [comment]
In this case you ought to be able to just call [`history.back()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/history), which will trigger a restoration visit.

**Comment by gregblass** [comment]
So in the end, using history.back() wasn't quite good enough in my case. What if the user navigates directly to a page with the back button? Or from a page outside of the app to a page with a back button? Or after a user submits a form? Or in the case that an infinite navigation loop is present? All of these cases are difficult to solve (or cannot be solved) using just history.back().

There is nothing quite like having the navigation stack handy for your back logic to make native-style back navigation work. It is the best solution and obviously why you guys implemented it.

In the end, since I don't have access to restore from the navigation stack on the web, my solution ended up being passing url parameters to override back paths in cases where the navigation flow was out of the ordinary/special (I.E., coming from an activity feed). That at least gave 1 level of back path override functionality - and beyond that, if the user goes a little farther down the nav path, I'm OK with it getting reset to the norm. I don't love it, and it dirties up my URL bars, but its good enough for now.

I would encourage opening up the nav stack restore to the web if possible?

**Comment by gregblass** [comment]
Here's the solution that I came up with that works great as far as I can tell:

```
<script>
    function goBack() {
      var backPath = "<%= @back_path %>";
      history.back();
      Turbolinks.visit(backPath, { action: "replace" });
    }
  </script>
```

The back path fallback goes to a statically defined path as a fallback. The action:replace prevents loops.

**Comment by gregblass** [comment]
It still doesn't prevent the back button from leaving the app if the previous page was another site. Its probably a pretty slim use case, but I wish I could say "if the previous page in the history was not in the app, use the static back path fallback." Apparently this is not possible in the browser due to security concerns.

**Comment by elsurudo** [comment]
@sstephenson My use case is that I have a "new" action, and after successful form submission, I'd like to do a restoration visit (I take it I can do `history.back()` here as you have suggested), but also update the flash on that _previous_ page. Any tips on how I can achieve this?

**Comment by elsurudo** [comment]
Nevermind, the flash takes care of it :)

**Comment by elsurudo** [comment]
However, the flash doesn't display when caching is enabled (makes sense). Any workaround for this?

**Comment by woahdae** [comment]
@sstephenson Sorry to at-you, not sure if I need a new ticket or just a follow up on this one.

OP asks, and I'll +1:

    I'm sure there was a reason to not allow restoration visits - just wondering what it was?

My use case, FWIW, is implementing breadcrumbs. With that UI element, the user can go further back in history than just one level. One might argue to use `window.history.go(numPages)`, but our particular breadcrumbs don't necessarily follow the user's browsing history. The crumbs follow a happy path that for many users will be their rough history, but is actually a sitemap-style breadcrumb, so not necessarily. Some pages, in particular search, are both a bit heavy in loading and the user will have various elements scrolled, panned, or selected, so it's pretty cool to just load it from cache and have them right back to where they were (with a bit of scroll position fiddling before render).

Since I already took the time to make it work well when the user navigated back, it was actually ridiculously simple. I just gave the breadcrumb links `"data-turbolinks-action"="restore"` and it all just worked! Yay.

Except I'm using an internal feature. Why?
